listener.go

```diff
@@ -208,10 +208,25 @@ func (l *Listener) consume() {
 
 	counter := 0
 	timer := time.Now()
+	needsRefreshIterator := false
 
 	for {
+		if !l.shouldConsume() {
+			l.setConsuming(false)
+			break
+		}
+
 		l.throttle(&counter, &timer)
 
+		if needsRefreshIterator {
```

Comment:
Move this inline with the loop (and `throttle`), in hopes of reducing the number of `ProvisionedThroughputExceededExceptions`